### PR TITLE
[SPARK-37865][SQL]Spark should not dedup the grouping Expressions when the first child of union has duplicate columns

### PR DESCRIPTION
**Pull request by chasingegg**
### What changes were proposed in this pull request?

When the first child of Union has duplicate columns like select a, a from t1 union select a, b from t2, spark only use the first column to aggregate the results, which would make the results incorrect, and this behavior is inconsistent with other engines like PostgreSQL, MySQL. **We could alias the attribute of the first child of union to resolve this, or you could argue that this is the feature of Spark SQL**.

sample query:
select
a,
a
from values (1, 1), (1, 2) as t1(a, b)
UNION ALL
SELECT
c,
d
from values (2, 3), (2, 3) as t2(c, d)

result is   (1, 1), (1, 1), (3, 3), (3, 3) 
expected (1, 1), (1, 1), (2, 3), (2, 3)

---

select
a,
a
from values (1, 1), (1, 2) as t1(a, b)
UNION
SELECT
c,
d
from values (2, 3), (2, 3) as t2(c, d)

result is  (1, 1), (2, 2)
expected (1, 1), (2, 3)

### Why are the changes needed?

It is possibly a bug.

### Does this PR introduce _any_ user-facing change?
 
Yes. When we union with the first child has duplicate columns, the result would be different.

### How was this patch tested?

Add new UT.

### CODE REVIEW

**Comment by LantaoJin** [comment]
> Does this PR introduce any user-facing change?
> Yes. When we union with the first child has duplicate columns, the result would be different.

@chasingegg I think this PR doesn't introduce any user-facing change. The Spark behavior which this PR to fix is buggy.

**Comment by LantaoJin** [comment]
Jenkins, ok to test

**Comment by LantaoJin** [comment]
cc @wangyum @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveUnion.scala [hunk 1]
```diff
@@ -208,6 +208,17 @@ object ResolveUnion extends Rule[LogicalPlan] {
     _.containsPattern(UNION), ruleId) {
     case e if !e.childrenResolved => e
 
+    case Union(children, false, false)
+        // the first child of Union has duplicate columns
+        if children.head.output.distinct.length < children.head.output.length =>
```
Comment (LantaoJin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveUnion.scala:213]: How about `select a, A` in a case-sensitive or case-insensitive analyzer?

**Comment by cloud-fan** [comment]
Yes this is a bug. Can you explain more about how this bug happens? Maybe adding alias is not the best way to fix it.

**Comment by chasingegg** [comment]
> Yes this is a bug. Can you explain more about how this bug happens? Maybe adding alias is not the best way to fix it.

@cloud-fan The way of adding alias is to regard the same attribute as different exprId, Presto did the same way, so I use the solution. I have spent some time debugging the behavior why it would cause problems while the first child of union has  duplicate attributes.
One part is in the optimization rule called PushProjectionThroughUnion,
image

In the buildReWrites, it will construct a map between the first child of union to the second child, for example,  the first child has duplicate columns like **a, a**, it is completely the same, and the second child fetch **c and d** column, it will generate the map with **a -> d**，making the second child's outputs are all the same as **d column**.
But After I remove the rule, the behavior is that the second child's outputs are all the same as **c column**... I would need to take another look.

**Comment by chasingegg** [comment]
@cloud-fan Hello, Besides the PushProjectionThroughUnion as I explained above, I guess the code below causes this bug,
```
  private def collectFromPlan(plan: SparkPlan): Array[T] = {
    val fromRow = resolvedEnc.createDeserializer()
    plan.executeCollect().map(fromRow)
  }
```
This `fromRow` method would bindReference for deserialization, in the second child of Union, it will take the first column twice as the result because the first child of Union has duplicate columns, and the first child determines the final union result's attributes.
If you have time, could you please take a look, thx~

**Comment by cloud-fan** [comment]
@chasingegg thanks for the explanation! When does Spark start to have this bug? Can you try to run the example SQL in different Spark versions?

This is a correctness bug and we usually need to backport, if the fix is not very complicated.

**Comment by LantaoJin** [comment]
@chasingegg 
```
select
a,
a
from values (1, 1), (1, 2) as t1(a, b)
UNION ALL
SELECT
c,
d
from values (2, 3), (2, 3) as t2(c, d)

result is (1, 1), (1, 1), (3, 3), (3, 3)
expected (1, 1), (1, 1), (2, 3), (2, 3)
```
I cannot reproduce the result of the SQL in community spark. Can you double-check it?

**Comment by LantaoJin** [comment]
Ok, I tested the first SQL in bin/spark-sql. There is no problem in 2.3/3.0/3.2. But it can be reproduced in bin/spark-shell, whatever in 2.3/3.0/3.0. And the second SQL can reproduce both in spark-shell and spark-sql from 2.3 to 3.2. cc @cloud-fan

**Comment by LantaoJin** [comment]
retest it please

**Comment by LantaoJin** [comment]
Retest this please

**Comment by LantaoJin** [comment]
@chasingegg please allow github actions and re-sync your forked branch to the master

**Comment by chasingegg** [comment]
> I think a complete fix is to let `Union` has fresh attribute IDs for its output, as technically `Union`'s outputs are totally different from its first child's outputs. It's going to be a big change so we should only do it in the master branch.
> 
> Can we open a PR for 3.0 with this surgical fix? Let's make sure the test still passes with this fix in 3.0.

Actually now we haven't passed the all tests in master branch. In pyspark module, it seems that it has `repeat` function which makes that two columns are the same like `a, a`, and `select a` works fine, but after this fix it would complained that `it is ambigous`, so I'm wondering if we could remove the pyspark thing or carefully disable the alias operation when the `original union` works fine. WDYT @cloud-fan

**Comment by LantaoJin** [comment]
> > I think a complete fix is to let `Union` has fresh attribute IDs for its output, as technically `Union`'s outputs are totally different from its first child's outputs. It's going to be a big change so we should only do it in the master branch.
> > Can we open a PR for 3.0 with this surgical fix? Let's make sure the test still passes with this fix in 3.0.
> 
> Actually now we haven't passed the all tests in master branch. In pyspark module, it seems that it has `repeat` function which makes that two columns are the same like `a, a`, and `select a` works fine, but after this fix it would complained that `it is ambigous`, so I'm wondering if we could remove the pyspark thing or carefully disable the alias operation when the `original union` works fine. WDYT @cloud-fan

I think this test failure only occurs in branch-3.2 and above since pandas. @chasingegg you can try to file a PR to 3.0.

**Comment by chasingegg** [comment]
Create the PR to 3.0 #35290.

**Pull request closed without merging**